src/app/main.cpp

```diff
@@ -362,6 +403,94 @@ void displayBadArgMessage(const QString &message)
 #endif
 }
 
+namespace
+{
+#ifndef Q_OS_WIN
+/*
+ * Borrowed from https://stackoverflow.com/a/758054
+ *
+ * getexename - Get the filename of the currently running executable
+ *
+ * The getexename() function copies an absolute filename of the currently
+ * running executable to the array pointed to by buf, which is of length size.
+ *
+ * If the filename would require a buffer longer than size elements, NULL is
+ * returned, and errno is set to ERANGE; an application should check for this
+ * error, and allocate a larger buffer if necessary.
+ *
+ * Return value:
+ * NULL on failure, with errno set accordingly, and buf on success. The
+ * contents of the array pointed to by buf are undefined on error.
+ *
+ * Notes:
+ * This function is tested on Linux only. It relies on information supplied by
+ * the /proc file system.
+ * The returned filename points to the final executable loaded by the execve()
+ * system call. In the case of scripts, the filename points to the script
+ * handler, not to the script.
+ * The filename returned points to the actual exectuable and not a symlink.
+ *
+ */
+char* getexename(char *buf, std::size_t size)
+{
+#ifdef NO_PROC_SELF_DIR
+    char linkname[64]; // /proc/<pid>/exe
+
+    // Get our PID and build the name of the link in /proc
+    pid_t pid = ::getpid();
+
+    if (snprintf(linkname, sizeof(linkname), "/proc/%i/exe", pid) < 0)
+    {
+        // This should only happen on large word systems. I'm not sure
+        // what the proper response is here.
+        // Since it really is an assert-like condition, aborting the
+        // program seems to be in order.
+        ::abort();
+    }
+
+    // Now read the symbolic link
+    ssize_t ret = readlink(linkname, buf, size);
+#else
+    ssize_t ret = ::readlink("/proc/self/exe", buf, size);
+#endif
+
+    // In case of an error, leave the handling up to the caller
+    if (ret == -1)
+        return nullptr;
+
+    // Report insufficient buffer size
+    if (static_cast<std::size_t>(ret) >= size) {
+        errno = ERANGE;
+        return nullptr;
+    }
+
+    // Ensure proper null-termination
+    buf[ret] = 0;
+    return buf;
+}
+#endif
+}
+
+QString executablePath()
+{
+#ifdef Q_OS_WIN
+    WCHAR apath[MAX_PATH];
+    ::GetModuleFileNameW(NULL, apath, MAX_PATH);
+    return QString::fromWCharArray(apath);
+#else
+    std::unique_ptr<char[]> buf = nullptr;
+    for (std::size_t size = 32; buf.reset(new char[size]), !getexename(buf.get(), size); size *= 2) {
+        // Anything but ERANGE indicates a real error
+        if (errno != ERANGE) {
+            qDebug() << "getexename() failed with error " << errno;
+            return {};
+        }
+    }
+
+    return QString::fromLocal8Bit(buf.get());
```

Comment:
Is Linux uses 8 bit encoding for filenames?

Reply:
Usually UTF-8.

Reply:
Excuse me, don't understand the question.

Reply:
Excuse me, I don't understand the question.

Reply:
You said the filenames are usually utf-8 encoded but you read it via fromLocal8Bit instead of fromUtf8... is it correct? Or am I missing something?

Reply:
I'm not sure UTF-8 qualifies as a 8-bit encoding but fromLocal8Bit() works fine in this case (fromLocalMultiByte() would be a better name).

Reply:
>but fromLocal8Bit() works fine in this case

As long as the filename doesn't contain any non-latin1 character, I suppose.

Reply:
> As long as the filename doesn't contain any non-latin1 character, I suppose.

I suppose it uses mbstowcs() and as such works with UTF-8.

Reply:
You must be exhausted. Otherwise, you could not confuse the 8-bit encoding with UTF-8 (some kind of Unicode).
fromLocal8Bit() loads 8-bit characters and decode them using some code page (e.g. cp1251, koi8-r, etc.).
As long as the test filename contains only latin-1 characters it can be equally successfully decoded by any of fromLatin1(), fromLocal8Bit() and fromUtf8(). But when string contains some multibyte utf-8 character it will be decoded as some number of characters if you apply fromLocal8Bit() to it.
Try the test code:
```c++
const char buf[] = "Строка с нелатинскими символами";
const QString str1 = QString::fromUtf8(buf);
qDebug() << str1; // will print "Строка с нелатинскими символами"
const QString str2 = QString::fromLocal8Bit(buf);
qDebug() << str2; // will print some kind of bullshit
```

Reply:
You are wrong here. I already wrote that the Qt's function name is confusing. It actually takes a multi-byte string in the current locale.

P.S.  And the sample outputs correct Cyrillic string two times.

Reply:
>From QString document, it seems fromLocal8Bit will depend on OS locale/codepage, right?

Well, after reading several levels of Qt docs I've suppose it's correct to use this method. But it really have confusing name. It should be "fromLocalEncoding()" since it can decode not only 8-bit strings but also Unicode.
Since all others should use the same locale on the same system it should work correctly. But fromUtf8() will fails on systems with non-utf8 locale settings.
(All above is not applicable to Windows).